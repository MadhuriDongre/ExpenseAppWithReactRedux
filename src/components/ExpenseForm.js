import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css'

const now = moment().format('MMM Do, YYYY');
console.log(now)

export default class ExpenseForm extends React.Component {
    constructor(props){
        super();
        this.state={
            description:props.expense?props.expense.description:'',
            note:props.expense?props.expense.note:'',
            amount:props.expense?props.expense.amount.toString():'',
            createdAt: props.expense?moment(props.expense.createdAt):moment(),
            focused:false,
            error:null
        };
    }
    onDescriptionChange=(e)=>{
        const description = e.target.value;
        this.setState(()=>({description}));
    };

    onNoteChange=(e)=>{
        const note = e.target.value;
        this.setState(()=>({note}));
    };
    onAmountChange=(e)=>{
        const amount = e.target.value;
        (!amount ||amount.match(/^\d{1,}(\.\d{0,2})?$/))? this.setState(()=>({amount})):null;
    }
    onDateChange=(createdAt)=>{
        // const createdAt=e.target.value
        createdAt?this.setState(()=>({createdAt})):null
    }
    onFocusChange=({focused})=>{
        this.setState(()=>({focused}))
    }
    onSubmit=(e)=>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            const error= "Please provide Description and Amount to create expense"
            this.setState(()=>({error}))
        }
        else{
            this.setState(()=>({error:null}))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10),
                note:this.state.note,
                createdAt:this.state.createdAt.valueOf()
            })
        }
    }
    render(){
        return (
            <div>
                <p>ExpenseForm text</p>
                <form onSubmit={this.onSubmit}> 
                    <input type="text" placeholder="description" value={this.state.description} autoFocus
                        onChange={this.onDescriptionChange}
                    />
                    <input type="text" value ={this.state.amount} placeholder="amount"
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={this.onDateChange} // PropTypes.func.isRequired
                        focused={this.state.focused} // PropTypes.bool
                        onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                        numberOfMonths={1}
                        isOutsideRange={()=>false}
                    />
                    <textarea placeholder="add a note for expense (optional)" value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button>Add Expense</button>
                    {this.state.error && <p className="warning">{this.state.error}</p>}
                </form>
            </div>
        );
    }
}