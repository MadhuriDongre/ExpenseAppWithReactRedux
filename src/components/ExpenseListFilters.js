import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate,setEndDate,setStartDate} from '../actions/filters';
import { DateRangePicker } from 'react-dates'

class ExpenseListFilters extends React.Component{
    state={
        focused:null
    }
    onDatesChange=({startDate, endDate})=>{
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }
    onFocusChange=(focused)=>{
        this.setState(()=>({focused}))
    }   
    render(){
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={(e)=>{
                    this.props.dispatch(setTextFilter(e.target.value));
                }}/>
                <select value={this.props.filters.sortBy} onChange={(e)=>{
                    e.target.value==="amount"?this.props.dispatch(sortByAmount(e.target.value)):this.props.dispatch(sortByDate(e.target.value));
                }}>
                    <option value ="date">Date</option>
                    <option value ="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                />
            </div>
        )
    }
}
// const ExpenseListFilters =(props)=>(
//     <div>
//         <input type="text" value={props.filters.text} onChange={(e)=>{
//             props.dispatch(setTextFilter(e.target.value));
//         }}/>
//         <select value={props.filters.sortBy} onChange={(e)=>{
//             e.target.value==="amount"?props.dispatch(sortByAmount(e.target.value)):props.dispatch(sortByDate(e.target.value));
//         }}>
//             <option value ="date">Date</option>
//             <option value ="amount">Amount</option>
//         </select>

//     </div>
// );

const mapStateToProps=(state)=>{
    return {
        filters:state.filters
    }
}
export default connect(mapStateToProps)(ExpenseListFilters);