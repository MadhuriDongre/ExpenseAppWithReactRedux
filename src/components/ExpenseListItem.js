import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral'

const ExpenseListItem =({description,amount,createdAt,id})=>(
    <div>
        <h2><Link to={`/edit/${id}`}>{description}</Link></h2>
        <p> {<span>&#x20B9;</span>}{ numeral(amount).format('0,0[.]00') } -- {moment(createdAt).format('MMMM Do, YYYY')}</p>
    </div>
);

export default ExpenseListItem;