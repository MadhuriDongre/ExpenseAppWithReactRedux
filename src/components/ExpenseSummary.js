import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral'
import selectExpenseTotal from '../selectors/expenses-total';
import selectExpense from '../selectors/expenses';

export const ExpenseSummary =({expenseCount, expensesTotal})=>(
    <div>
        <p>{`Viewing ${expenseCount} expense(s) totalling `}{<span>&#x20B9;</span>}{` ${numeral(expensesTotal).format('0,0[.]00')}`}</p>
    </div>
);

const mapStateToProps = (state)=>{
    const visibleExpense=selectExpense(state.expenses,state.filters)
    return {
        expenseCount:visibleExpense.length,
        expensesTotal:selectExpenseTotal(visibleExpense)
    }
};

export default connect(mapStateToProps)(ExpenseSummary);