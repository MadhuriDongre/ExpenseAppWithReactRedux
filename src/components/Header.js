import React from 'react';
import { NavLink} from 'react-router-dom';
import { connect} from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header =({startLogout})=>(
    <header>
        <h1>Expensify</h1>
            <NavLink activeClassName="isActive" to="/dashboard" exact>Dashboard</NavLink>
            <NavLink activeClassName="isActive" to="/create">Create Expense</NavLink>
            <button onClick={startLogout}>LogOut</button>
    </header>
);

const mapDispatchToProps=(dispatch)=>({
    startLogout:()=>dispatch(startLogout())
})
export default connect(undefined,mapDispatchToProps)(Header);