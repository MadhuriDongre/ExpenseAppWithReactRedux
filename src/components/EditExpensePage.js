import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense} from '../actions/expenses'

const EditExpensePage =(props)=>{
    console.log(props);
    return (
        <div>
           <ExpenseForm 
                onSubmit={(expense)=>{
                    props.dispatch(editExpense(props.expenses.id,expense));
                    props.history.push('/');
                }}
                expense={props.expenses}
            />
            <button onClick={()=>{
                props.dispatch(removeExpense({id:props.expenses.id}));
                props.history.push('/')
            }}>Remove</button>
        </div>
    );
};

const mapStateToProps=(state,props)=>{
    return {
        expenses:state.expenses.find((expense)=>expense.id === props.match.params.id)
    }
}
export default connect(mapStateToProps)(EditExpensePage);