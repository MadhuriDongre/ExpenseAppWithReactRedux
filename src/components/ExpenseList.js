import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpense from '../selectors/expenses';

export const ExpenseList =(props)=>(
    <div>
        <p>ExpenseList</p>
        {props.expenses.length===0?(<p>No expense found!!</p>): (props.expenses  && props.expenses.map((expense)=>{
            return (
                <ExpenseListItem key={expense.id} {...expense}/>
            )}))
        }
    </div>
);

const mapStateToProps = (state)=>{
    return {
        expenses:selectExpense(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseList);