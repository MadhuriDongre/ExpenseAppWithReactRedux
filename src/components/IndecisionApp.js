import React from 'react';
import AddOption from './AddOption';
import Options from './Options';
import Header from './Header';
import Action from './Action';
import OptionModal from './OptionModal';

export default class IndecisionApp extends React.Component{
    state ={
        options : [],
        selectedOption : undefined
    }
    componentDidMount(){
        try {
            const json = localStorage.getItem('options');
            const options = JSON.parse(json);
            if(options){
                this.setState(()=>({options}));
            }
        }
        catch(e){
            //swallow the error
        }
    }
    componentDidUpdate(prevProps, prevState){
        if(prevState.options.length!=this.state.options.length){
            const json = JSON.stringify(this.state.options);
            localStorage.setItem('options',json);
        }
    }
    componentWillUnmount(){
        console.log('componentWilUnmount');
    }
    handleDeleteAll= ()=>{
        // this.setState(()=>{
        //     return {
        //         options:[]
        //     };
        // });

        this.setState( ()=>({ options:[] })  );
    }

    handleDeleteOption=(removeOption)=>{
        this.setState((prevState)=>({ options: prevState.options.filter((opt)=>opt!==removeOption) }));
    }
    handlePick=()=>{
        if(this.state.options.length>0){
            let option = Math.floor(Math.random()*this.state.options.length);
            this.setState(()=>({selectedOption: this.state.options[option]}));
        }
    }

    handleAddOption=(option)=>{
        if(!option){
            return 'Please enter valid Option';
        }
        else if(this.state.options.indexOf(option)>-1){
            return 'This Option already exist!';
        }
        this.setState((prevState)=>({options: prevState.options.concat(option)}));
    }

    handleClearOption = ()=>{
        this.setState(()=>({selectedOption: undefined}));
    }
    render(){
        const subtitle ="Put your life in the Hands of Computer!!";

        return (
            <div>
                <Header subtitle={subtitle}/>
                <div className="container">
                    <Action hasOption ={this.state.options.length>0} handlePick= {this.handlePick}/>
                    <div className="widget">
                        <Options  options={this.state.options} handleDeleteAll={this.handleDeleteAll} handleDeleteOption={this.handleDeleteOption}/>
                        <AddOption handleAddOption={this.handleAddOption}/>
                    </div>
                </div>
                <OptionModal selectedOption= {this.state.selectedOption} handleClearOption = {this.handleClearOption}/>
            </div>
        );
    }
}