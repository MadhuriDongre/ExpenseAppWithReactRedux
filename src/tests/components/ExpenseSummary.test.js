import React from 'react';
import { shallow } from 'enzyme';
import {ExpenseSummary} from '../../components/ExpenseSummary';

test('should render expense list with expenses',()=>{
    const wrapper = shallow(<ExpenseSummary expenseCount={2} expensesTotal={200}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render expense list without expenses',()=>{
    const wrapper = shallow(<ExpenseSummary expenseCount={0} expensesTotal={0}/>);
    expect(wrapper).toMatchSnapshot();
});