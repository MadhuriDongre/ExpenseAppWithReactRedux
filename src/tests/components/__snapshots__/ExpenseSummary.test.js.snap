// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list with expenses 1`] = `
<div>
  <p>
    Viewing 2 expense(s) totalling 
    <span>
      ₹
    </span>
     200
  </p>
</div>
`;

exports[`should render expense list without expenses 1`] = `
<div>
  <p>
    Viewing 0 expense(s) totalling 
    <span>
      ₹
    </span>
     0
  </p>
</div>
`;
